airflow/operators/python.py

```diff
@@ -185,16 +186,37 @@ def execute(self, context: Dict):
 class ShortCircuitOperator(PythonOperator, SkipMixin):
     """
     Allows a workflow to continue only if a condition is met. Otherwise, the
-    workflow "short-circuits" and downstream tasks are skipped.
+    workflow "short-circuits" and downstream tasks are skipped. The short-circuiting can be configured to
+    operated in one of two modes: "hard" or "soft".  In "hard" mode, the default mode, all downstream tasks
+    are skipped without considering the ``trigger_rule`` defined for tasks.  In "soft" mode, the
+    ``trigger_rule`` is respected.
 
     The ShortCircuitOperator is derived from the PythonOperator. It evaluates a
-    condition and short-circuits the workflow if the condition is False. Any
-    downstream tasks are marked with a state of "skipped". If the condition is
-    True, downstream tasks proceed as normal.
+    condition and short-circuits the workflow if the condition is False. Any downstream tasks are marked with
+    a state of "skipped" based on the short-circuiting mode configured. If the condition is True, downstream
+    tasks proceed as normal.
 
-    The condition is determined by the result of `python_callable`.
+    The condition is determined by the result of ``python_callable``.
+
+    :param mode: If set to "hard", all downstream tasks from this operator task will be skipped.  This is
+        the default behavior.  If set to "soft", downstream tasks will be skipped but the ``trigger_rule``
+        defined for a task will be respected.
+    :type mode: bool
     """
 
+    modes = ["hard", "soft"]
```

Comment:
I wonder if it would be better to have:
```suggestion
    downstream_trigger_rules_mode = ["respect", "ignore"]
```
At least for me i can understand from the parameter name and option what it means without looking it up in the docstring/documentation. WDYT?

Reply:
Maybe even `ignore_downstream_trigger_rules` and boolean typed? (Not sure what the opinion is for literal vs boolean param values is if there is one.) Makes the param actionable and maybe even easier to understand if True/False controlled. Definitely a case of explicit is better than implicit but the initial critique was the initial `do_hard_short` was a lengthy param name. @uranusjr WDYT?

Reply:
The biggest problem I have with `do_hard_short` was it's difficult to internalise (is it `do_soft_short=True` or `do_hard_short=False`?) so I think the same would still apply if we change this to a longer name. `downstream_trigger_mode="xxx"` is long, but not difficult to remember, so it's fine IMO. I'm less sure about the mode names though; from the original issue description, maybe something like `direct` and `all`, or `direct-only` and `recursive` would work better.

Reply:
What about `respect_trigger_rules = True/False` ?

Reply:
Perhaps. Airflow isn't that consistent with this sort of flags right now, but I guess `ignore_` is used more right now? (see e.g. `ignore_task_deps`)

Reply:
@uranusjr that was the basis for my suggestion of `ignore_downstream_trigger_rules`

Reply:
I think I'm OK with `ignore_downstream_trigger_rules` (a bit long but we have autosuggest in editors).

Reply:
@eladkal If you are OK with the name, I can go ahead and update the code.

Reply:
yep

Reply:
why not use the same semantics than for the sensors ( soft/hard )->

https://github.com/apache/airflow/pull/8867

soft_skip:bool = False

Reply:
@josh-fell ?

Reply:
In previous iterations we did have "soft/hard modes" and `do_hard_short` but the question was what do they _mean_? While something like `ignore_downstream_trigger_rules` is long, it is pretty explicit and more understandable than the `modes`/`do_hard_short` names. IMO the same goes with `soft_skip` (or `soft_fail` in the sensors). They are succinct parameter names but what do they _mean_?  Also, there is more consistency (albeit loosely) around `ignore_...` flag nomenclature in Airflow.

Even if users look at the documentation/docstrings anyway to figure out what the parameter does, I still think following the "explicit is better than implicit" principle is the way to go.

Reply:
Looks mich better now indeed Hard/Soft was extremely difficult to mentally map what it could mean.